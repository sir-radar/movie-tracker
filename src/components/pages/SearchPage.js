import React,{useEffect} from 'react';
import { connect } from 'react-redux';
import PageLayout from '../layout/PageLayout';
import SearchBox from '../SearchBox';
import MovieCard from '../MovieCard';
import Loader from '../Loader';
import NoData from '../NoData';
import { search } from '../../store/actions/searchActions';
import { getSessionID } from '../../store/actions/authActions';
import { addToFavourite, removeFromFavourite } from '../../store/actions/favouritesActions';
import { addToWatchList, removeFromWatchList } from '../../store/actions/watchListActions';

const SearchPage = (props) => {
  const { createSession,
          session_id, 
          searchStatus, 
          movies, 
          addToFavourite, 
          removeFavourite, 
          favouriteIDs, 
          search,
          removeFromWatchList,
          addToWatchList,
          watchlistIDs 
        } = props;

  //create session_id if it doesn't exist
  useEffect(() => {
    if(!session_id){
      createSession()
    }
  }, [createSession, session_id]);

  return (
        
        <PageLayout
        
          showNav = {true}

          searchbox = {
            <SearchBox
              onSearch={ (title) => search(title) }
            />
          }

          pageContent = {
            //display on successful request
            (searchStatus === 'SUCCESS') ? movies.results.map(result => <MovieCard
                key={result.id}
                id={result.id}
                title={result.title}
                image={result.poster_path}
                overview={result.overview}
                addToFavourite={addToFavourite}
                removeFavourite={removeFavourite}
                favouriteIDs={favouriteIDs}
                removeFromWatchList={removeFromWatchList}
                addToWatchList={addToWatchList}
                watchlistIDs={watchlistIDs}
              />) 
            
            : null
          }

          loader = {
            //display while request is pending
            (searchStatus === 'PENDING')
            ? <Loader/>
            : null
          }

          nodata = {
            //display if there is no result
            (searchStatus === 'SUCCESS' && movies.results.length < 1)
            ? <NoData/>
            : null
          }
          
        />
  );
}

const mapStateToProps = (state) => (
  {
    searchStatus: state.status.search,
    searchError: state.status.searchError,
    movies: state.search,
    session_id: state.auth.session_id,
    favouriteIDs: state.favourites.favouriteMoviesID,
    watchlistIDs: state.watchlists.watchListsID,
  }
);
  
const mapDispatchToProps = (dispatch) => (
  {
    search: (title) => dispatch(search(title)),
    createSession: () => dispatch(getSessionID()),
    addToFavourite: (payload) => dispatch(addToFavourite(payload)),
    removeFavourite: (payload) => dispatch(removeFromFavourite(payload)),
    addToWatchList: (payload) => dispatch(addToWatchList(payload)),
    removeFromWatchList: (payload) => dispatch(removeFromWatchList(payload)),
  }
)

export default connect(mapStateToProps, mapDispatchToProps) (SearchPage);
