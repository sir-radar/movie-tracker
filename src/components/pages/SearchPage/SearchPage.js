import React,{useState} from 'react';
import { connect } from 'react-redux';
import PageLayout from '../../layout/PageLayout';
import SearchBox from '../../SearchBox/SearchBox';
import MovieCard from '../../MovieCard/MovieCard';
import Loader from '../../Loader/Loader';
import NoData from '../../NoData/NoData';
import Error from '../../Error/Error';
import ErrorMessage from '../../ErrorMessage/ErrorMessage';
import Pagination from '../../Pagination/Pagination';
import { search } from '../../../store/actions/searchActions';
import { addToFavourite, removeFromFavourite } from '../../../store/actions/favouritesActions';
import { addToWatchList, removeFromWatchList } from '../../../store/actions/watchListActions';
import { resetStatus } from '../../../store/actions/statusActions';

const SearchPage = (props) => {

  const { searchStatus, 
          movies, 
          addToFavourite, 
          removeFavourite, 
          favouriteIDs, 
          search,
          removeFromWatchList,
          addToWatchList,
          watchlistIDs,
          watchlistAction,
          favouriteAction,
          resetStatus
        } = props;

  const [searchQuery, setSearchQuery] = useState('')
  console.log(watchlistAction)
  //gets searchquery for pagination
  const updateSerachQuery = (value) => {
    setSearchQuery(value)
  }

  const loadMore = (page) =>{
    search(searchQuery, page)
  }

  return (
        
        <PageLayout

          showNav = {true}

          searchbox = {
            <SearchBox
              onSearch={ (title) => search(title) }
              updateSerachQuery={(query) => updateSerachQuery(query)}
            />
          }

          pageContent = {
            //display on successful request
            (searchStatus === 'SUCCESS') ? movies.results.map(result => <MovieCard
                key={result.id}
                id={result.id}
                title={result.title}
                image={result.poster_path}
                overview={result.overview}
                addToFavourite={addToFavourite}
                removeFavourite={removeFavourite}
                favouriteIDs={favouriteIDs}
                removeFromWatchList={removeFromWatchList}
                addToWatchList={addToWatchList}
                watchlistIDs={watchlistIDs}
              />) 
            
            : null
          }

          loader = {
            //display while request is pending
            (searchStatus === 'PENDING')
            ? <Loader/>
            : null
          }

          error = {
            //display when an error occurs
            (searchStatus === 'ERROR')
            ? <Error/>
            : null
          }

          errorMessage = {
            //display when an error occurs in favouriting or adding movie to watchlist
            (watchlistAction === 'ERROR' || favouriteAction === 'ERROR' )
            ? <ErrorMessage resetStatus={resetStatus}/>
            : ''
          }

          nodata = {
            //display if there is no result
            (searchStatus === 'SUCCESS' && movies.results.length < 1)
            ? <NoData/>
            : null
          }

          pagination = {
            //pagination
            (searchStatus === 'SUCCESS') ? movies.total_results > 20 ? <Pagination 
                                          searchMore={ (page) => loadMore(page) }
                                          page={movies.page} 
                                          pages={movies.total_pages}
                                          /> 
                    : ''
                    : ''
          }
          
        />
  );
}

const mapStateToProps = (state) => (
  {
    searchStatus: state.status.search,
    watchlistAction: state.status.watchlistAction,
    favouriteAction: state.status.favouriteAction,
    movies: state.search,
    favouriteIDs: state.favourites.favouriteMoviesID,
    watchlistIDs: state.watchlists.watchListsID,
  }
);
  
const mapDispatchToProps = (dispatch) => (
  {
    search: (title, page) => dispatch(search(title, page)),
    addToFavourite: (payload) => dispatch(addToFavourite(payload)),
    removeFavourite: (payload) => dispatch(removeFromFavourite(payload)),
    addToWatchList: (payload) => dispatch(addToWatchList(payload)),
    removeFromWatchList: (payload) => dispatch(removeFromWatchList(payload)),
    resetStatus: () => dispatch(resetStatus()),
  }
)

export default connect(mapStateToProps, mapDispatchToProps) (SearchPage);
