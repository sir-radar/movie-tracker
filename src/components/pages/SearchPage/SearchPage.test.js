import React from 'react';
import { MemoryRouter } from "react-router-dom";
import { render, cleanup } from '@testing-library/react';
import SearchPage from './SearchPage';
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { Provider } from 'react-redux';
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";

//reducers
// import search from '../../../store/reducers/searchReducer';
// import status from '../../../store/reducers/statusReducer';
// import favourites from '../../../store/reducers/favouritesReducer';
// import watchlists from '../../../store/reducers/watchListReducer';

//actions
// import {searchRequest} from '../../../store/actions/searchActions';

const buildStore = configureStore([thunk]);

afterEach(cleanup)

const renderComponent = (states) => {
  const store = buildStore({...states});
  const div = document.createElement("div");
  return render(<Provider store={store}><MemoryRouter><SearchPage /></MemoryRouter></Provider>, div)
}

const renderComponentForSnapshot = (states) => {
  const store = buildStore({...states});
  return <Provider store={store}><MemoryRouter><SearchPage /></MemoryRouter></Provider>
}

describe('Search Page Component', () => {

  it("renders without crashing", () => {
    const states = {
      search:{},
      favourites:{},
      status:{},
      watchlists:{}
    }
    renderComponent(states)
  })

  it("renders nav component", () => {
    const states = {
      search:{},
      favourites:{},
      status:{},
      watchlists:{}
    }
    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("nav")).not.toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
  })

  it("renders search box component", () => {
    const states = {
      search:{},
      favourites:{},
      status:{},
      watchlists:{}
    }
    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("search-box")).not.toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
  })

  it("renders loading component on pending request", () => {
   
    const states = {
      search:{},
      favourites:{},
      status:{
        search: 'PENDING',
        favourite:'',
        favouriteAction:'',
        watchlist:'',
        watchlistAction:''
      },
      watchlists:{}
    }

    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("search-box")).not.toBeNull()
    expect(queryByTestId("loader")).not.toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()

  })

  it("renders Movie Card component on successful request", () => {
    const states = {
      search:{
        results:[
          {
            id: 1,
            title: "Some text",
            overview: 'some text',
            poster_path:""
          }
        ]
      },
      favourites:{},
      status:{
        search: 'SUCCESS',
        favourite:'',
        favouriteAction:'',
        watchlist:'',
        watchlistAction:''
      },
      watchlists:{},
      auth:{
        session_id: "GHHjsjskkslsllsghsjsksk"
      },
    }
    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("movie-card")).not.toBeNull()
    expect(queryByTestId("pagination")).toBeNull()

  })


  it("renders NoData component on successful request with empty data", () => {

    const states = {
      search:{
        results:[]
      },
      favourites:{},
      status:{
        search: 'SUCCESS',
        favourite:'',
        favouriteAction:'',
        watchlist:'',
        watchlistAction:''
      },
      watchlists:{},
      auth:{
        session_id: "GHHjsjskkslsllsghsjsksk"
      },
    }

   
    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
    expect(queryByTestId("nodata-component")).not.toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()

  })

  it("renders Error component on failed request", () => {

    const states = {
      search:{
        results:[]
      },
      favourites:{},
      status:{
        search: 'ERROR',
        favourite:'',
        favouriteAction:'',
        watchlist:'',
        watchlistAction:''
      },
      watchlists:{},
      auth:{
        session_id: "GHHjsjskkslsllsghsjsksk"
      },
    }
    const {queryByTestId} = renderComponent(states)
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).not.toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()

  })

  it("renders Pagination component on successful request with more than 20 results", () => {

    const states = {
      search:{
        results:[
          {id: 1, title:'', overview:''},
          {id: 2,title:'', overview:''}
        ],
        total_results: 21,
        page: 1,
        pages: 2
      },
      favourites:{},
      status:{
        search: 'SUCCESS',
        favourite:'',
        favouriteAction:'',
        watchlist:'',
        watchlistAction:''
      },
      watchlists:{},
      auth:{
        session_id: "GHHjsjskkslsllsghsjsksk"
      },
    }

    const {queryByTestId, queryAllByTestId} = renderComponent(states)
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryAllByTestId("movie-card").length).toBe(states.search.results.length)
    expect(queryByTestId("pagination")).not.toBeNull()

  })

  it("renders Error Message component on failed add/remove favourite or watchlist request", () => {

    const states = {
      search:{
        results:[
          {id: 1, title:'', overview:''},
          {id: 2,title:'', overview:''}
        ],
        total_results: 21,
        page: 1,
        pages: 2
      },
      favourites:{},
      status:{
        search: 'SUCCESS',
        favourite:'',
        favouriteAction:'ERROR',
        watchlist:'',
        watchlistAction:'ERROR'
      },
      watchlists:{},
      auth:{
        session_id: "GHHjsjskkslsllsghsjsksk"
      },
    }
    const {queryByTestId, queryAllByTestId} = renderComponent(states)
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("error-message")).not.toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryAllByTestId("movie-card").length).toBe(states.search.results.length)
    expect(queryByTestId("pagination")).not.toBeNull()

  })


  it("matches snapshot", () => {
    const states = {
      search:{},
      favourites:{},
      status:{},
      watchlists:{}
    }
    const tree = renderer.create(renderComponentForSnapshot(reducers)).toJSON();
    expect(tree).toMatchSnapshot();
  })

})