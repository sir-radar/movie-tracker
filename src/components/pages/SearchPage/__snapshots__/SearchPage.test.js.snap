// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Page Component matches snapshot 1`] = `
Array [
  <nav
    className="d-flex justify-content-center"
    data-testid="nav"
  >
    <a
      className="p-3 primary-text-color active"
      data-testid="search"
      href="/"
      onClick={[Function]}
    >
      Search
    </a>
    <a
      className="p-3 primary-text-color"
      data-testid="favourite"
      href="/favourites"
      onClick={[Function]}
    >
      Favorites
    </a>
    <a
      className="p-3 primary-text-color"
      data-testid="watchlater"
      href="/watchlist"
      onClick={[Function]}
    >
      WatchLater
    </a>
  </nav>,
  <main
    className="mt-4 p-3 col-10 col-md-10 offset-1 offset-md-1"
  >
    <div
      className="row justify-content-center"
      data-testid="search-box"
    >
      <div
        className="col-md-10"
      >
        <form
          data-testid="search-form"
          onSubmit={[Function]}
        >
          <div
            className="input-group mb-3"
          >
            <input
              className="form-control"
              data-testid="search-input"
              onChange={[Function]}
              placeholder="Search movie title..."
              type="text"
            />
            <div
              className="input-group-append"
            >
              <button
                className="btn px-3 btn-background"
              >
                <i
                  className="fa fa-search text-white"
                />
              </button>
            </div>
          </div>
          <p
            className="error text-danger"
            data-testid="search-error"
          >
            
          </p>
        </form>
      </div>
    </div>
    <div
      className="row"
    >
      
      
    </div>
  </main>,
]
`;
